bundles/org.openhab.binding.adorne/src/main/java/org/openhab/binding/adorne/internal/hub/AdorneHubController.java

```diff
@@ -0,0 +1,595 @@
+        // Note this is best effort. If we are unlucky the hub can still enter waiting on input just after our stop
+        // here. Because waiting on input is long-running we can't just synchronize it with the stop check as case 2
+        // above. But that is ok as waiting on input has a timeout and will honor stop after that.
+        Socket hubSocket = this.hubSocket;
+        if (hubSocket != null) {
+            try {
+                hubSocket.shutdownInput();
+            } catch (IOException e) {
+                logger.debug("Couldn't shutdown hub socket");
+            }
+        }
+    }
+
+    /**
+     * Stops the hub controller once all in-flight commands have been executed.
+     */
+    public void stopWhenCommandsServed() {
+        stopWhenCommandsServed = true;
+    }
+
+    /**
+     * Stops the hub controller after a certain point in time.
+     *
+     * @param stopTimestamp timestamp to stop after
+     */
+    public void stopBy(long stopTimestamp) {
+        this.stopTimestamp = stopTimestamp;
+    }
+
+    /**
+     * Turns device on or off.
+     *
+     * @param zoneId the device's zone ID
+     * @param on true to turn on the device
+     */
+    public void setOnOff(int zoneId, boolean on) {
+        sendRestCmd(String.format(HUB_REST_SET_ONOFF, getNextCommandId(), zoneId, on));
+    }
+
+    /**
+     * Sets the brightness for a device. Applies only to dimmer devices.
+     *
+     * @param zoneId the device's zone ID
+     * @param level A value from 1-100. Note that in particular value 0 is not supported, which means this method can't
+     *            be used to turn off a dimmer.
+     */
+    public void setBrightness(int zoneId, int level) {
+        if (level < 1 || level > 100) {
+            throw new IllegalArgumentException();
+        }
+        sendRestCmd(String.format(HUB_REST_SET_BRIGHTNESS, getNextCommandId(), zoneId, level));
+    }
+
+    /**
+     * Gets asynchronously the state for a device.
+     *
+     * @param zoneId the device's zone ID
+     * @return a future for the {@link AdorneDeviceState}
+     */
+    public CompletableFuture<AdorneDeviceState> getState(int zoneId) {
+        sendRestCmd(String.format(HUB_REST_REQUEST_STATE, getNextCommandId(), zoneId));
+        synchronized (stateCommandsLock) {
+            CompletableFuture<AdorneDeviceState> stateCommand = stateCommands.get(zoneId);
+            if (stateCommand != null) {
+                return (stateCommand);
+            } else {
+                stateCommand = new CompletableFuture<>();
+                stateCommands.put(zoneId, stateCommand);
+                return (stateCommand);
+            }
+        }
+    }
+
+    /**
+     * Gets asynchronously all zone IDs that are in use on the hub.
+     *
+     * @return a future for the list of zone IDs
+     */
+    public CompletableFuture<List<Integer>> getZones() {
+        sendRestCmd(String.format(HUB_REST_REQUEST_ZONES, getNextCommandId()));
```

Comment:
Why are you already sending a rest command if a completable future for this command might already exist?

Reply:
See above